Game logic

pipe2442 opened the pull request:
Milestone 3: Game Logic

In this milestone we:
- Added logic to our game.
- Created method to follow game rules.
- Created a turn control flow.
- Added error validations for correct game flow.

pipe2442 committed:
Created game_logic branch and board.rb file

pipe2442 committed:
Started  game logic dev

pipe2442 committed:
Added feature to deply game board and pick users  turn

pipe2442 committed:
Added functionality and correct linters

pipe2442 committed:
Added game logic and corrections for gaming flow

pipe2442 committed:
Fixed sopt taken bug

pipe2442 committed:
Fixed initial bugs for game_logic and win validations

pipe2442 committed:
Fixed bucle control flow and linter with correct  logic

pipe2442 committed:
Merge branch 'development' into game_logic

pipe2442 committed:
Fixed linters

pipe2442 committed:
Fixed linters

pipe2442 committed:
Merge branch 'game_logic' of https://github.com/pipe2442/Tic-Tac-Toe-Ruby into game_logic

pipe2442 committed:
Fixed linters

pipe2442 committed:
Fix linter errors with empty guarclause space

pipe2442 committed:
Fix linter errors with empty guarclause space

enaburekhan reviewed (commented):
# Invalid
* Friendly reminder: You should make sure that your pull request looks professional before review. Please take a step back and read a lesson here https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243642/. Then apply 3 simple rules from this lesson (it will be easy as you can use a template provided there). This Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready!
* 
You can reach me on slack @ enaburekhan for further clarification.

Happy Codind!!!

pipe2442 committed:
added readme file template

pipe2442 commented:
Hi @enaburekhan thanks for your advice, the comment on the  pull request was updated and the correct readme file template was added to the project, thanks for your advice!

GzimAsani commented on lib/control.rb:
```diff
@@ -0,0 +1,58 @@
+end
+
+def draw_case
+  puts 'SORRY, ITS A DRAW NO WINNER' if @turn == 9 && @winner == false
```
You should not use puts outside of bin/main.rb file
![attr_accessor](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

GzimAsani commented on lib/control.rb:
```diff
@@ -0,0 +1,58 @@
+    @valid_move = false
+    @turn -= 1
+  elsif (@board[player_sel - 1] == 'X') || (@board[player_sel - 1] == 'O')
+    puts 'THAT SPOT IS ALREADY TAKEN'
```
You should not use puts outside of bin/main.rb file
![attr_accessor](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

GzimAsani commented on lib/control.rb:
```diff
@@ -0,0 +1,58 @@
+
+def valid_moves(player_sel, _board)
+  if (player_sel < 1) || (player_sel > 9)
+    puts 'THATS NOT A VALID MOVE!'
```
You should not use puts outside of bin/main.rb file
![attr_accessor](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

GzimAsani commented on lib/control.rb:
```diff
@@ -0,0 +1,58 @@
+
+def winner_message
+  if @turn.odd?
+    puts "Congrats #{@first_player} you are the winner"
```
You should not use puts outside of bin/main.rb file
![attr_accessor](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

GzimAsani commented on lib/board.rb:
```diff
@@ -0,0 +1,42 @@
+      puts "#{player} please select a move based on the board numbers:"
+      puts ' '
+      create_board
+      player_sel = gets.chomp.to_i
```
You should not use gets.chomp outside of bin/main.rb file
![attr_accessor](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

GzimAsani commented on lib/board.rb:
```diff
@@ -0,0 +1,42 @@
+        selection = 'O'
+      end
+      puts "#{player} please select a move based on the board numbers:"
+      puts ' '
```
You should not use print outside of bin/main.rb file
![attr_accessor](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

GzimAsani commented on lib/board.rb:
```diff
@@ -0,0 +1,42 @@
+    puts '-----------'
+    puts " #{@board[3]} | #{@board[4]} | #{@board[5]} "
+    puts '-----------'
+    puts " #{@board[6]} | #{@board[7]} | #{@board[8]} "
```
You should not use puts outside of bin/main.rb file
![attr_accessor](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/only_main_print.png)

GzimAsani reviewed (changes requested):
### CHANGES REQUIRED  :stop_sign: 
### HELLO @pipe2442 You did a great job in this millstone :+1:  very nice work :+1:  but there are some changes that you have to make, please check below: 
- [ ] Please be reminded that  `puts/print` and `gets.chomp`  should only be inside bin/main.rb
- [ ] You should remove inline comments left after the previous milestone
- [ ] Your screenshot picture on README file is now showing, kindly fix this

### if you have any question kindly reach me on slack @GzimAsani

### HAPPY CODING :computer:

pipe2442 committed:
added puts and prints just to main - fixed linters - removed comments

pipe2442 committed:
Uploaded project image

Uploaded project image for readme.txt file.

pipe2442 committed:
Update README.md

Added project image to readme file.

pipe2442 commented:
Hi, @GzimAsani thanks for your review! I removed comments, added puts and prints just to the main file, and fixed the readme image, thanks for your advice!

pipe2442 committed:
Update README.md

Updated readme file with correct project information.

rindrajosia commented on bin/main:
```diff
@@ -1,52 +1,87 @@
-  puts 'Its players turn'
-  puts 'Please select an available move based on the board numbers '
-  puts ' '
+def draw_case
```
Please, put this logic inside the lib directory like your win_validator

rindrajosia commented on bin/main:
```diff
@@ -1,52 +1,87 @@
-    # HERE WE CREATE CONDITION TO CHECK FOR A DRAW
-    puts 'Its a draw'
-    break
+class GameBoard
```
Class should be in the lib folder.

rindrajosia reviewed (changes requested):
**Status: Changes Required**
Very good job! Overall, this project is awesome! You are VERY close to finishing it! 
We only require you to change some minor issues in order to get your project approved:
- [ ]  The project uses at least two classes (Usually it uses three: Game, Board, and Player).
- [ ] The code of the project should be logically divided between classes. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone3/classes.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))
- [ ] All game-logic related resources (classes and/or modules) are kept in the lib directory. ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/OOP/milestone1/lib_dir_exist.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59565018/))

After addressing those minor bugs submit another CRR.

Happy Coding!
Rindra Josia

pipe2442 committed:
Removed logic code lib - gave prints and puts just to main class - fixed linters

pipe2442 commented:
Thanks for your advice @rindrajosia I fixed the issues that you requested on the project 👍

pipe2442 committed:
Added move_selection method to print different players turn

pipe2442 committed:
Fix linters

NtwaliHeritier reviewed (approved):
**Approved ✔**
Hello Team, 
I can tell you put in a lot of work. Merge the branches and move on to the next step.

[OPTIONAL] Implement a class Player that holds the logic of the player, like name, keys, etc
[OPTIONAL] Display the name of the player on each move.
[OPTIONAL] Put the game logic codes in game class not player.

These are optional changes if you want to make the game. However you have my blessings to merge and continue with another milestone. 👏

Happy coding.
